clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -209,48 +228,76 @@ object GarbageCollector {
   ): Dataset[Row] = {
     val commitsDF = getCommitsDF(runID, commitDFLocation, spark)
     val rangesDF = getRangesDFFromCommits(commitsDF, repo, conf)
-    getExpiredEntriesFromRanges(rangesDF, conf, repo)
+    val expired = getExpiredEntriesFromRanges(rangesDF, conf, repo)
+
+    // validate deduplications
+    val activeRangesDF = rangesDF.where("!expired")
+    subtractDeduplications(expired, activeRangesDF, conf, repo, spark)
+  }
+
+  private def subtractDeduplications(expired: Dataset[Row], activeRangesDF: Dataset[Row], conf: APIConfigurations, repo: String, spark: SparkSession) : Dataset[Row]= {
+    val expiredAddr: Set[String] = expired.select("address").collect().map(_.getString(0)).toSet
+    val ranges : Seq[String] = activeRangesDF.select("range_id").collect().map(_.getString(0)).toSeq.distinct
+    val rangesRDD = spark.sparkContext.parallelize(ranges)
+
+    val activeAddresses =rangesRDD.flatMap(range=> {
+      getEntryAddressesIfInSet(range, conf, repo, expiredAddr)
+    }).map(x => Row(x))
```

Comment:
This looks like a flatMap with side data.  I *think* that it will be somewhat tricky, because `expiredAddr` needs to be broadcast to all the workers.  It looks like it is some kind of JOIN, so I would prefer if you could write `ranges` as an RDD (or any other Spark table, I don't really know the difference...) and join with that here.